File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/API/Maps/ImageGenerator.java
package jp.aoichaan0513.A_TosoGame_Live.API.Maps;

import java.awt.*;
import java.awt.geom.AffineTransform;
import java.awt.image.BufferedImage;
import java.util.List;

public class ImageGenerator {

    public static BufferedImage getRotatedImage(int angle, BufferedImage img) {
        int width = 0;
        int height = 0;
        // 元画像の読み込み
        BufferedImage sourceImage = null;
        BufferedImage outputImage = null;
        sourceImage = img;
        width = sourceImage.getWidth();
        height = sourceImage.getHeight();

        if (angle % 180 != 0) {
            outputImage = new BufferedImage(height, width, BufferedImage.TYPE_INT_RGB);
        } else {
            // 180度の倍数の時は画像サイズの変更は行わない。
            outputImage = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        }
        Graphics2D out = outputImage.createGraphics();
        AffineTransform at = new AffineTransform();

        if (angle == 0) {
            // 何もしない
        } else if (angle == 180) {
            at.setToRotation(Math.toRadians(-1.0 * angle), width / 2, height / 2);
        } else if (angle == 270) {
            at.setToRotation(Math.toRadians(-1.0 * angle), width / 2, height / 2);
            at.translate((width - height) / 2, (width - height) / 2);
        } else {
            at.setToRotation(Math.toRadians(-1.0 * angle), width / 2, height / 2);
            at.translate(-(width - height) / 2, -(width - height) / 2);
        }
        out.drawImage(sourceImage, at, null);
        return outputImage;
    }

    public static BufferedImage getPaintedImage(int width, int height, List<String> command) {
        int size = height < width ? width : height;
        int addx = 0;
        int addy = 0;
        if (width != height) {
            if (width < height) {
                addx = (height - width) / 2;
            } else {
                addy = (width - height) / 2;
            }
        }
        BufferedImage im = new BufferedImage(size, size, BufferedImage.TYPE_INT_ARGB);
        Graphics g = im.getGraphics();
        g.setColor(new Color(206, 180, 136));
        g.fillRect(0, 0, im.getWidth(), im.getHeight());

        for (String cmd : command) {
            String[] data = cmd.split(",");
            /*
             * [DataFormat]
             * 0: x
             * 1: y
             * 2: r
             * 3: g
             * 4: b
             */
            g.setColor(new Color(Integer.parseInt(data[2]), Integer.parseInt(data[3]), Integer.parseInt(data[4])));
            g.drawLine(Integer.parseInt(data[0]) + addx, Integer.parseInt(data[1]) + addy, Integer.parseInt(data[0]) + addx, Integer.parseInt(data[1]) + addy);
        }

        im.flush();
        g.dispose();
        return im;
    }

    public static BufferedImage getResizedImage(int size, BufferedImage img) {
        Image tmp = img.getScaledInstance(size, size, Image.SCALE_DEFAULT);
        BufferedImage dimg = new BufferedImage(size, size, BufferedImage.TYPE_INT_ARGB);

        Graphics2D g2d = dimg.createGraphics();
        g2d.drawImage(tmp, 0, 0, null);
        g2d.dispose();

        return dimg;
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/API/Map/ImageGenerator.kt
package jp.aoichaan0513.A_TosoGame_Live.API.Map

import jp.aoichaan0513.A_TosoGame_Live.Utils.DataClass
import java.awt.Color
import java.awt.Image
import java.awt.geom.AffineTransform
import java.awt.image.BufferedImage

class ImageGenerator {
    companion object {

        fun getRotatedImage(angle: Int, img: BufferedImage): BufferedImage {
            val sourceImage = img

            val width = sourceImage.width
            val height = sourceImage.height

            val outputImage = if (angle % 180 != 0) {
                BufferedImage(height, width, BufferedImage.TYPE_INT_RGB)
            } else {
                BufferedImage(width, height, BufferedImage.TYPE_INT_RGB)
            }

            val out = outputImage.createGraphics()
            val at = AffineTransform()
            if (angle == 0) {
                // 何もしない
            } else if (angle == 180) {
                at.setToRotation(Math.toRadians(-1.0 * angle), width / 2.toDouble(), height / 2.toDouble())
            } else if (angle == 270) {
                at.setToRotation(Math.toRadians(-1.0 * angle), width / 2.toDouble(), height / 2.toDouble())
                at.translate((width - height) / 2.toDouble(), (width - height) / 2.toDouble())
            } else {
                at.setToRotation(Math.toRadians(-1.0 * angle), width / 2.toDouble(), height / 2.toDouble())
                at.translate(-(width - height) / 2.toDouble(), -(width - height) / 2.toDouble())
            }
            out.drawImage(sourceImage, at, null)
            return outputImage
        }

        fun getPaintedImage(width: Int, height: Int, list: List<DataClass.BlockData>): BufferedImage {
            val size = if (height < width) width else height

            var addX = 0
            var addY = 0

            if (width != height) {
                if (width < height)
                    addX = (height - width) / 2
                else
                    addY = (width - height) / 2
            }

            val im = BufferedImage(size, size, BufferedImage.TYPE_INT_ARGB)

            val graphics = im.graphics
            graphics.color = Color(206, 180, 136)
            graphics.fillRect(0, 0, im.width, im.height)

            for (blockData in list) {
                val (x, y) = Pair(blockData.x + addX, blockData.y + addY)
                val (r, g, b) = blockData.rgbColor

                graphics.color = Color(r, g, b)
                graphics.drawLine(x, y, x, y)
            }

            im.flush()
            graphics.dispose()
            return im
        }

        fun getResizedImage(size: Int, img: BufferedImage): BufferedImage {
            val tmp = img.getScaledInstance(size, size, Image.SCALE_DEFAULT)
            val bufferedImage = BufferedImage(size, size, BufferedImage.TYPE_INT_ARGB)
            val graphics2D = bufferedImage.createGraphics()
            graphics2D.drawImage(tmp, 0, 0, null)
            graphics2D.dispose()
            return bufferedImage
        }
    }
}